libraries/client/client_converter_callback.py

```diff
@@ -0,0 +1,350 @@
+            App.logger.debug('Multiple project, part {0} of {1}, converting book {2}'.
+                             format(part_id, part_count, book))
+            multiple_project = True
+        else:
+            part_count = None
+            part_id = None
+            multiple_project = False
+
+        self.job.success = self.success
+        for message in self.log:
+            self.job.log_message(message)
+        for message in self.warnings:
+            self.job.warnings_message(message)
+        for message in self.errors:
+            self.job.error_message(message)
+        if len(self.errors):
+            self.job.log_message('{0} function returned with errors.'.format(self.job.convert_module))
+        elif len(self.warnings):
+            self.job.log_message('{0} function returned with warnings.'.format(self.job.convert_module))
+        else:
+            self.job.log_message('{0} function returned successfully.'.format(self.job.convert_module))
+
+        self.job.update()
+
+        s3_commit_key = 'u/{0}'.format(self.job.identifier)
+        upload_key = s3_commit_key
+        if multiple_project:
+            upload_key += "/" + part_id
+
+        App.logger.debug('Callback for commit {0}...'.format(s3_commit_key))
+
+        # Download the ZIP file of the converted files
+        converted_zip_url = self.job.output
+        converted_zip_file = os.path.join(self.temp_dir, converted_zip_url.rpartition('/')[2])
+        remove(converted_zip_file)  # make sure old file not present
+        download_success = True
+        App.logger.debug('Downloading converted zip file from {0}...'.format(converted_zip_url))
+        try:
+            download_file(converted_zip_url, converted_zip_file)
+        except:
+            download_success = False  # if multiple project we note fail and move on
+            if not multiple_project:
+                remove_tree(self.temp_dir)  # cleanup
+            if self.job.errors is None:
+                self.job.errors = []
+            self.job.errors.append("Missing converted file: " + converted_zip_url)
+        finally:
+            App.logger.debug('download finished, success={0}'.format(str(download_success)))
+
+        if download_success:
+            # Unzip the archive
+            unzip_dir = self.unzip_converted_files(converted_zip_file)
+
+            # Upload all files to the cdn_bucket with the key of <user>/<repo_name>/<commit> of the repo
+            self.upload_converted_files(upload_key, unzip_dir)
+
+        if multiple_project:
+            # Now download the existing build_log.json file, update it and upload it back to S3
+            build_log_json = self.update_build_log(s3_commit_key, part_id + "/")
+
+            # mark part as finished
+            self.cdn_upload_contents({}, s3_commit_key + '/' + part_id + '/finished')
+
+            # check if all parts are present, if not return
+            missing_parts = []
```

Comment:
This section needs to be removed and replaced with call to ClientLinterCallback.deploy_if_conversion_finished()

Reply:
PS - I also need to document my function

Reply:
I suppose we need to save the final merged build log  back to jobs table.  I need to add that to my function.